### DISCUSSION TITLE
Using `lonboard` in a Databricks notebook

### DISCUSSION THREAD

**Discussion by jpd-defra** [developmentseed/lonboard#144] [Q&A] [answered]
Hello,

Thanks for sharing `lonboard`, I'm very excited at the prospect of using it. 

I currently use Databricks for my python work. I have tried producing an interactive map with the following:

![db_eg](https://github.com/developmentseed/lonboard/assets/106590727/dd3566f3-8daf-4180-ac9b-22ce5e179b10)

While the execution doesn't fail, the widget/map doesn't display. I've read that people have got round [similar issues with folium](https://stackoverflow.com/questions/55648077/folium-map-not-showing-datbricks-python) by exporting to html then displaying. Has anyone got an idea how I could do this with `lonboard` please? 

_Note: My cluster is a above the minimum runtime version suggested, and I have tried running multiple times._

**Comment by kylebarron** [comment, marked as answer]
Per their docs, [some custom widgets are supported](https://docs.databricks.com/en/notebooks/ipywidgets.html#which-third-party-jupyter-widgets-are-supported-in-databricks), but IIRC Databricks doesn't support the full Jupyter Widgets protocol. You can look in your browser console to see if there's any errors, and you might want to try rendering a map with an empty/very small dataset to make sure it's not related to the size of your data. I don't have access to a Databricks instance, and so I don't have a way to test on Databricks.

Edit: Actually if I had to guess, the issue is probably with Databricks requiring an older version of JavaScript than what we're using. We're building on top of [Anywidget](https://github.com/manzt/anywidget), which makes it 100x easier to build Jupyter Widgets, but compiles to a recent version of JS. We could try transpiling our JS bundle to ES5 but I don't think that would work with anywidget.

https://github.com/developmentseed/lonboard/issues/8 is tracking export to HTML, but that's a decent chunk of work and I don't have a timeline suggestion.

**Reply by jpd-defra** [reply]
Thanks for the swift response Kyle, that's all very helpful insight. 

I checked again with one small, simple polygon and the same issue arose so I don't think it was a dataset size issue. Databricks can be very frustrating at times. I'm going to look for a way to utilise a Jupyter notebook for dataviz in my organisation (potentially publishing with Posit Connect). I'll keep an eye out for HTML developments in future though! 

Thanks again.

**Comment by kylebarron** [comment]
Version 0.3 added export to a static HTML file, so that may be compatible with databricks

**Reply by jpd-defra** [reply]
Thanks for letting me know. I've given it a go, and unfortunately still unsuccessful. Here's my code and output if it's of any interest, but as you say, this is almost certainly a Databricks issue!

```
from lonboard import SolidPolygonLayer, Map
from geopandas import read_parquet

dat = read_parquet(f'/Workspace/Repos/{username}/sds-15min-greenspace/data/focal_boxes')
layer = SolidPolygonLayer.from_geopandas(dat[['geometry']])
mymap = Map(layers = [layer])
mymap.to_html(f'/dbfs/mnt/lab/unrestricted/{username}/sandbox/lonboard_test.html')
```

When trying to use `displayHTML` the map doesn't render and only displays the path:

![image](https://github.com/developmentseed/lonboard/assets/106590727/e9649234-2937-42cc-9a0c-27c885218f01)

Trying an alternative with `IPython`:

![image](https://github.com/developmentseed/lonboard/assets/106590727/ac939410-0956-4194-a854-00604541782d)

Clicking the output to show the error reveals:

![image](https://github.com/developmentseed/lonboard/assets/106590727/9db46441-868f-4f37-b6c4-4fd7f65049f3)

**Reply by kylebarron** [reply]
`SyntaxError: Invalid or unexpected token` means that databricks requires an older version of JS, because of the way databricks notebooks are set up, which I don't think is feasible for us to try and support.

**Reply by jpd-defra** [reply]
No problem - I appreciate you letting me know about the html export updates. I have a jupyter notebook alternative plan so hoping I can make use of lonboard for my work.

**Comment by danielsparing** [comment]
It currently works when calling `m.as_html()`.
